core/src/main/java/com/linecorp/armeria/common/metric/MetricKey.java

```diff
@@ -0,0 +1,321 @@
+     */
+    public MetricKey withLabel(MetricLabel label, String value) {
+        validateLabel(label, value);
+        final ImmutableMap<MetricLabel, String> newLabels = ImmutableMap.<MetricLabel, String>builder()
+                .putAll(labels).put(label, value).build();
+        return new MetricKey(nameParts, newLabels);
+    }
+
+    /**
+     * Returns a newly created {@link MetricKey} which has the specified labels added.
+     */
+    public MetricKey withLabels(Map<?, String> labels) {
+        requireNonNull(labels, "labels");
+        if (labels.isEmpty()) {
+            return this;
+        }
+
+        final ImmutableMap.Builder<MetricLabel, String> builder = ImmutableMap.builder();
+        if (!this.labels.isEmpty()) {
+            builder.putAll(this.labels);
+        }
+
+        labels.forEach((k, v) -> {
+            if (k instanceof MetricLabel) {
+                builder.put((MetricLabel) k, v);
+            } else if (k instanceof CharSequence) {
+                builder.put(new MetricLabelImpl(k.toString()), v);
+            } else {
+                throw new IllegalArgumentException(
+                        "labels contains a key of unsupported type: " + k.getClass().getName() +
+                        " (expected: " + MetricLabel.class.getSimpleName() +
+                        " or " + CharSequence.class.getSimpleName() + ')');
+            }
+        });
+
+        return new MetricKey(nameParts, builder.build());
+    }
+
+    static void validateNamePart(String namePart) {
+        requireNonNull(namePart, "namePart");
+        checkArgument(!namePart.isEmpty(), "empty namePart");
+    }
+
+    static void validateLabel(String label, String value) {
+        requireNonNull(label, "label");
+        checkArgument(!label.isEmpty(), "empty label");
+        requireNonNull(value, "value");
+    }
+
+    static void validateLabel(MetricLabel label, String value) {
+        requireNonNull(label, "label");
+        checkArgument(!label.name().isEmpty(), "empty label name");
+        requireNonNull(value, "value");
+    }
+
+    @Override
+    public int hashCode() {
+        if (hashCode != 0) {
+            return hashCode;
+        }
+
+        return hashCode = nameParts.hashCode() * 31 + labels.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (!(o instanceof MetricKey)) {
+            return false;
+        }
+
+        final MetricKey that = (MetricKey) o;
+        return hashCode() == that.hashCode() &&
+               nameParts.equals(that.nameParts) &&
+               labels.equals(that.labels);
+    }
+
+    @Override
+    public String toString() {
+        final String name = String.join(".", nameParts);
+        if (labels.isEmpty()) {
+            return name;
+        }
+
+        // We use Joiner here because labels.toString() inserts space after comma (,).
```

Comment:
Is there a difference between `Joiner` and `String.join(",", labels.entrySet())`? We use `String.join` just above.